src/storage/src/sink/iceberg.rs

```diff
@@ -0,0 +1,1413 @@
+                                            .context("Failed to write row to DeltaWriter")?;
+                                        written = true;
+                                        break;
+                                    }
+                                }
+                                if !written {
+                                    // ...otherwise stash it for later
+                                    let entry = stashed_rows.entry(row_ts).or_default();
+                                    entry.push((row, diff_pair));
+                                }
+                            }
+                        }
+                        Event::Progress(frontier) => {
+                            input_frontier = frontier;
+                        }
+                    }
+                }
+
+                // Check if frontiers have advanced, which may unlock batches ready to close
+                if PartialOrder::less_equal(
+                    &processed_batch_description_frontier,
+                    &batch_description_frontier,
+                ) || PartialOrder::less_than(&processed_input_frontier, &input_frontier)
+                {
+                    // Close batches whose upper is now in the past
+                    let mut ready_batches: Vec<_> = in_flight_batches
+                        .extract_if(|(lower, upper), _| {
+                            PartialOrder::less_than(lower, &batch_description_frontier)
+                                && PartialOrder::less_than(upper, &input_frontier)
+                        })
+                        .collect();
+
+                    // Sort by upper bound to process in timestamp order
+                    ready_batches.sort_by(|(a, _), (b, _)| {
+                        a.1.elements()
+                            .into_element()
+                            .cmp(b.1.elements().into_element())
+                    });
+
+                    if !ready_batches.is_empty() {
+                        let mut max_upper = Antichain::from_elem(Timestamp::minimum());
+                        for (desc, mut delta_writer) in ready_batches {
+                            let data_files = delta_writer
+                                .close()
+                                .await
+                                .context("Failed to close DeltaWriter")?;
+                            let mut data_files = data_files
+                                .into_iter()
+                                .map(|data_file| {
+                                    BoundedDataFile::new(
+                                        data_file,
+                                        current_schema.as_ref().clone(),
+                                        desc.clone(),
+                                    )
+                                })
+                                .collect::<Vec<_>>();
+
+                            output.give_container(&capset[0], &mut data_files);
```

Comment:
We generally avoid manually creating containers and let timely choose the appropriate size using its container builders. I suggest you turn the collected iteartor into a normal loop where you call `output.give()` for each data file